wallet/key_deriver_test.go

```diff
@@ -0,0 +1,266 @@
+		derivedPublicKey, err := keyDeriver.DerivePublicKey(
+			protocolID,
+			keyID,
+			wallet.WalletCounterparty{
+				Type:         wallet.CounterpartyTypeOther,
+				Counterparty: counterpartyPublicKey,
+			},
+			false,
+		)
+		assert.NoError(t, err)
+		assert.IsType(t, &ec.PublicKey{}, derivedPublicKey)
+	})
+
+	t.Run("should derive the correct public key for self", func(t *testing.T) {
+		derivedPublicKey, err := keyDeriver.DerivePublicKey(
+			protocolID,
+			keyID,
+			wallet.WalletCounterparty{
+				Type:         wallet.CounterpartyTypeOther,
+				Counterparty: counterpartyPublicKey,
+			},
+			true,
+		)
+		assert.NoError(t, err)
+		assert.IsType(t, &ec.PublicKey{}, derivedPublicKey)
+	})
+
+	t.Run("should derive the correct private key", func(t *testing.T) {
+		derivedPrivateKey, err := keyDeriver.DerivePrivateKey(
+			protocolID,
+			keyID,
+			wallet.WalletCounterparty{
+				Type:         wallet.CounterpartyTypeOther,
+				Counterparty: counterpartyPublicKey,
+			},
+		)
+		assert.NoError(t, err)
+		assert.IsType(t, &ec.PrivateKey{}, derivedPrivateKey)
+	})
+
+	t.Run("should derive the correct symmetric key", func(t *testing.T) {
+		derivedSymmetricKey, err := keyDeriver.DeriveSymmetricKey(
+			protocolID,
+			keyID,
+			wallet.WalletCounterparty{
+				Type:         wallet.CounterpartyTypeOther,
+				Counterparty: counterpartyPublicKey,
+			},
+		)
+		assert.NoError(t, err)
+		assert.NotEmpty(t, derivedSymmetricKey)
+	})
+
+	t.Run("should be able to derive symmetric key with anyone", func(t *testing.T) {
+		_, err := keyDeriver.DeriveSymmetricKey(
+			protocolID,
+			keyID,
+			wallet.WalletCounterparty{
+				Type: wallet.CounterpartyTypeAnyone,
+			},
+		)
+		assert.NoError(t, err)
+	})
+
+	t.Run("should reveal the correct counterparty shared secret", func(t *testing.T) {
+		sharedSecret, err := keyDeriver.DeriveSymmetricKey(
+			protocolID,
+			keyID,
+			wallet.WalletCounterparty{
+				Type:         wallet.CounterpartyTypeOther,
+				Counterparty: counterpartyPublicKey,
+			},
+		)
+		assert.NoError(t, err)
+		assert.NotEmpty(t, sharedSecret)
+	})
+
+	t.Run("should not reveal shared secret for self", func(t *testing.T) {
+		_, err := keyDeriver.DeriveSymmetricKey(
+			protocolID,
+			keyID,
+			wallet.WalletCounterparty{
+				Type: wallet.CounterpartyTypeSelf,
+			},
+		)
+		assert.Error(t, err)
+		assert.EqualError(t, err, "cannot derive symmetric key for self")
+	})
+
+	t.Run("should reveal the specific key association", func(t *testing.T) {
+		sharedSecret, err := keyDeriver.DeriveSymmetricKey(
+			protocolID,
+			keyID,
+			wallet.WalletCounterparty{
+				Type:         wallet.CounterpartyTypeOther,
+				Counterparty: counterpartyPublicKey,
+			},
+		)
+		assert.NoError(t, err)
+		assert.NotEmpty(t, sharedSecret)
+	})
+
+	t.Run("should throw an error for invalid protocol names", func(t *testing.T) {
+		testCases := []struct {
+			name        string
+			protocol    wallet.WalletProtocol
+			keyID       string
+			expectError bool
```

Comment:
Minor thing, But it was a little bit confusing that, 
the test name suggests that it should always return an error, 
given that, actually the exepctError flag, seems to be redundant (especially as it is always true)
and the code responsible for assertions based on this flag is also redundant.

In other words, by removing this flag, the test should be much simpler and clearer.

Reply:
Good catch.